cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,365 @@
+	mu       *sync.RWMutex // protects firewall.
+	firewall *godo.Firewall
+}
+
+// firewallManagerOp manages the interaction with the DO Firewalls API.
+type firewallManagerOp struct {
+	client             *godo.Client
+	fwCache            firewallCache
+	workerFirewallName string
+	workerFirewallTags []string
+}
+
+// FirewallManager retrieves and stores firewall representations.
+type FirewallManager interface {
+	// Get returns the current public access firewall representation (i.e., the DO Firewall object)
+	// if it exists and nil if it does not exist.
+	Get(ctx context.Context) (*godo.Firewall, error)
+
+	// Set applies the given inbound rules to the public access firewall.
+	Set(ctx context.Context, firewallRequest *godo.FirewallRequest) error
+}
+
+// FirewallController helps to keep cloud provider service firewalls in sync.
+type FirewallController struct {
+	kubeClient         clientset.Interface
+	client             *godo.Client
+	workerFirewallTags []string
+	workerFirewallName string
+	serviceLister      corelisters.ServiceLister
+	fwManager          FirewallManager
+}
+
+// NewFirewallController returns a new firewall controller to reconcile public access firewall state.
+func NewFirewallController(
+	ctx context.Context,
+	kubeClient clientset.Interface,
+	client *godo.Client,
+	serviceInformer coreinformers.ServiceInformer,
+	fwManager FirewallManager,
+	workerFirewallTags []string,
+	workerFirewallName string,
+) *FirewallController {
+	fc := &FirewallController{
+		kubeClient:         kubeClient,
+		client:             client,
+		workerFirewallTags: workerFirewallTags,
+		workerFirewallName: workerFirewallName,
+		fwManager:          fwManager,
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
+				defer cancel()
+				err := fc.ensureReconciledFirewall(ctx)
+				if err != nil {
+					klog.Errorf("failed to reconcile worker firewall: %s", err)
+				}
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
+				defer cancel()
+				err := fc.ensureReconciledFirewall(ctx)
+				if err != nil {
+					klog.Errorf("failed to reconcile worker firewall: %s", err)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
+				defer cancel()
+				err := fc.ensureReconciledFirewall(ctx)
+				if err != nil {
+					klog.Errorf("failed to reconcile worker firewall: %s", err)
+				}
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}, fm *firewallManagerOp, fwReconcileFrequency time.Duration) {
+	wait.Until(func() {
+		fc.actualRun(stopCh, fm, fwReconcileFrequency)
```

Comment:
We could now make `actualRun` return an `error` and log that here. This would possibly help with testing error cases, and generally makes for a cleaner approach to log at the most top-level only.